WordPress/gutenberg
# Theme export: convert uploaded URLs to relative path assets

🚧 

> [!WARNING]
> Experimental code - please don't treat it as ready yet. Thank you!



## What? How?

This is a first pass, so it's pretty rough and tumble. No tests or defensive coding yet.

This PR checks a theme tree for URIs containing the current site's upload path. If found:

- it replaces the paths with a relative equivalent: `'http://domain/wp-content/uploads/1/2/test.gif' >' file:./uploads/test.gif'`
- fetches the file and adds it to the zip in `/uploads`

The outcome is that the uploaded files become part of the new theme's asset library, and can therefore be installed anywhere.

## Why?

A theme can use uploaded files for:

- fonts
- background images

The paths to these assets are absolute URIs to the uploaded files.

Currently, when exporting themes these URIs are preserved in the theme.json, which means that the theme isn't "portable" in that relies on those hosted assets.


## Testing Instructions

1. In the site editor, go to global styles and upload a font or two.
2. Upload some images and use them as background images for the site or for quote/pullquote/group blocks.
3. Save the styles.
4. Now export the theme and open the zip.
5. In theme.json, the absolute paths to uploaded files should be converted to relative `file:./` paths
6. The uploaded files themselves should be available in `/uploads` in the zip.
7. Now upload the exported theme and activate it in your site. Check that it all works as expected.

### Exported theme.json [BEFORE]

```json
					"fontFace": [
						{
							"fontFamily": "\"Star Jedi\"",
							"fontStyle": "normal",
							"fontWeight": "400",
							"src": "http://localhost:8888/wp-content/uploads/fonts/Starjedi.ttf"
						}
					],
```

### Exported theme.json [AFTER]

```json
					"fontFace": [
						{
							"fontFamily": "\"Star Jedi\"",
							"fontStyle": "normal",
							"fontWeight": "400",
							"src": "file:./uploads/Starjedi.ttf"
						}
					],
```

## Screenshots or screencast 
<img width="282" alt="Screenshot 2024-10-25 at 7 20 17 am" src="https://github.com/user-attachments/assets/c1a4a3b6-b0fc-44ea-be19-3981d4536ecb">
